include/deal.II/base/polynomials_bernardi_raugel.h

```diff
@@ -0,0 +1,210 @@
+ * the face numbering in GeometryInfo. The vector $\mathbf{n}_i$ points in
+ * the positive axis direction and not necessarily normal to the element for
+ * consistent orientation across edges.
+ *<dl>
+ *   <dt> 2D bubble functions (in order)
+ *   <dd> $x=0$ edge: $\mathbf{p}_1 = \mathbf{n}_1 (1-x)(y)(1-y)$
+ *
+ *        $x=1$ edge: $\mathbf{p}_2 = \mathbf{n}_2 (x)(y)(1-y)$
+ *
+ *        $y=0$ edge: $\mathbf{p}_3 = \mathbf{n}_3 (x)(1-x)(1-y)$
+ *
+ *        $y=1$ edge: $\mathbf{p}_4 = \mathbf{n}_4 (x)(1-x)(y)$
+ *
+ *   <dt> 3D bubble functions (in order)
+ *   <dd> $x=0$ edge: $\mathbf{p}_1 = \mathbf{n}_1 (1-x)(y)(1-y)(z)(1-z)$
+ *
+ *        $x=1$ edge: $\mathbf{p}_2 = \mathbf{n}_2 (x)(y)(1-y)(z)(1-z)$
+ *
+ *        $y=0$ edge: $\mathbf{p}_3 = \mathbf{n}_3 (x)(1-x)(1-y)(z)(1-z)$
+ *
+ *        $y=1$ edge: $\mathbf{p}_4 = \mathbf{n}_4 (x)(1-x)(y)(z)(1-z)$
+ *
+ *        $z=0$ edge: $\mathbf{p}_5 = \mathbf{n}_5 (x)(1-x)(y)(1-y)(1-z)$
+ *
+ *        $z=1$ edge: $\mathbf{p}_6 = \mathbf{n}_6 (x)(1-x)(y)(1-y)(z)$
+ *
+ *</dl>
+ *
+ * Then the $BR_1(E)$ polynomials are defined on quadrilaterals and hexahedra
+ * by $BR_1(E) = Q_1(E) \oplus \mbox{span}\{\mathbf{p}_i, i=1,...,2d\}$.
+ *
+ *
+ * @ingroup Polynomials
+ * @author Graham Harper
+ * @date 2018
+ */
+template <int dim>
+class PolynomialsBernardiRaugel
+{
+public:
+  /**
+   * Constructor. Creates all basis functions for Bernardi-Raugel polynomials
+   * of given degree.
+   *
+   * @arg k: the degree of the Bernardi-Raugel-space, which is currently
+   * limited to the case <tt>k=1</tt>
+   */
+  PolynomialsBernardiRaugel(const unsigned int k);
+
+  /**
+   * Return the number of Bernardi-Raugel polynomials.
+   */
+  unsigned int
+  n() const;
+
+
+  /**
+   * Return the degree of Bernardi-Raugel polynomials.
+   * Since the bubble functions are quadratic in at least one variable,
+   * the degree of the Bernardi-Raugel polynomials is two.
+   */
+  unsigned int
+  degree() const;
+
+  /**
+   * Return the name of the space, which is <tt>BernardiRaugel</tt>.
+   */
+  std::string
+  name() const;
+
+  /**
+   * Compute the value and derivatives of each Bernardi-
+   * Raugel polynomial at @p unit_point.
+   *
+   * The size of the vectors must either be zero or equal <tt>n()</tt>.  In
+   * the first case, the function will not compute these values.
+   *
+   * If you need values or derivatives of all tensor product polynomials then
+   * use this function, rather than using any of the <tt>compute_value</tt>,
+   * <tt>compute_grad</tt> or <tt>compute_grad_grad</tt> functions, see below,
+   * in a loop over all tensor product polynomials.
+   */
+  void
+  compute(const Point<dim> &           unit_point,
+          std::vector<Tensor<1, dim>> &values,
+          std::vector<Tensor<2, dim>> &grads,
+          std::vector<Tensor<3, dim>> &grad_grads,
+          std::vector<Tensor<4, dim>> &third_derivatives,
+          std::vector<Tensor<5, dim>> &fourth_derivatives) const;
+  /**
```

Comment:
Add a space above this line

Reply:
A newline between lines 135 and 136, correct? Got it